cloudinit/cmd/tests/test_query.py

```diff
@@ -1,195 +1,267 @@
-        self.add_patch('cloudinit.cmd.query.read_cfg_paths', 'm_paths')
-        self.m_paths.return_value = paths
-        with mock.patch('sys.stderr', new_callable=StringIO) as m_stderr:
+        paths, run_dir, _, _ = self._setup_paths(tmpdir)
+        with mock.patch('cloudinit.cmd.query.read_cfg_paths') as m_paths:
+            m_paths.return_value = paths
             with mock.patch('os.getuid') as m_getuid:
                 m_getuid.return_value = 0
-                self.assertEqual(1, query.handle_args('anyname', args))
-        json_file = os.path.join(run_dir, INSTANCE_JSON_FILE)
-        sensitive_file = os.path.join(run_dir, INSTANCE_JSON_SENSITIVE_FILE)
-        self.assertIn(
-            'WARNING: Missing root-readable %s. Using redacted %s instead.' % (
-                sensitive_file, json_file),
-            m_stderr.getvalue())
+                assert 1 == query.handle_args('anyname', args)
+        json_file = run_dir.join(INSTANCE_JSON_FILE)
+        sensitive_file = run_dir.join(INSTANCE_JSON_SENSITIVE_FILE)
+        msg = (
+            'Missing root-readable %s. Using redacted %s instead.' %
+            (
+                sensitive_file.strpath, json_file.strpath
+            )
+        )
+        assert msg in caplog.text
 
-    def test_handle_args_root_uses_instance_sensitive_data(self):
-        """When no instance_data argument, root uses semsitive json."""
-        user_data = self.tmp_path('user-data', dir=self.tmp)
-        vendor_data = self.tmp_path('vendor-data', dir=self.tmp)
-        write_file(user_data, 'ud')
-        write_file(vendor_data, 'vd')
-        run_dir = self.tmp_path('run_dir', dir=self.tmp)
-        sensitive_file = os.path.join(run_dir, INSTANCE_JSON_SENSITIVE_FILE)
-        write_file(sensitive_file, '{"my-var": "it worked"}')
-        ensure_dir(run_dir)
-        paths = Paths({'run_dir': run_dir})
-        self.add_patch('cloudinit.cmd.query.read_cfg_paths', 'm_paths')
-        self.m_paths.return_value = paths
+    @pytest.mark.parametrize(
+        'ud_src,ud_expected,vd_src,vd_expected',
+        (
+            ('hi mom', 'hi mom', 'hi pops', 'hi pops'),
+            ('ud'.encode('utf-8'), 'ud', 'vd'.encode('utf-8'), 'vd'),
+            (_gzip_data(b'ud'), 'ud', _gzip_data(b'vd'), 'vd'),
+            (_gzip_data('ud'.encode('utf-8')), 'ud', _gzip_data(b'vd'), 'vd'),
+            (_gzip_data(b'ud') + b'invalid', 'ci-b64:',
+             _gzip_data(b'vd') + b'invalid', 'ci-b64:'),
+            # non-utf-8 encodable content
+            ('hi mom'.encode('utf-16'), 'ci-b64:',
+             'hi pops'.encode('utf-16'), 'ci-b64:'),
+        )
+    )
+    def test_handle_args_root_processes_user_data(
+        self, ud_src, ud_expected, vd_src, vd_expected, capsys, tmpdir
+    ):
+        """Support reading multiple user-data file content types"""
+        paths, run_dir, user_data, vendor_data = self._setup_paths(
+            tmpdir, ud_val=ud_src, vd_val=vd_src
+        )
```

Comment:
And a similar change could be made here (and in the following test):

```suggestion
    def test_handle_args_root_processes_user_data(
        self, ud_src, ud_expected, vd_src, vd_expected, capsys, paths, tmpdir
    ):
        """Support reading multiple user-data file content types"""
        user_data, vendor_data = self._setup_paths(ud_val=ud_src, vd_val=vd_src)
```

It follows that `self._setup_paths()` can be simplified to (a) only return `user_data`/`vendor_data`, and (b) require both of those arguments to be passed in.

Reply:
I think in this case it makes things a bit more complex to shift to paths as I don't get to use the tmpdir instance methods to write multiple files etc in individual tests so it looks to bloat my current use case here.

Reply:
> I think in this case it makes things a bit more complex to shift to paths as I don't get to use the tmpdir instance methods to write multiple files etc in individual tests so it looks to bloat my current use case here.

Hmm, I'm not sure I follow this, but I would like the `paths` fixture to be usable in situations like this. If you can still remember the context here, I'd like to understand what the gap for using it is.

(FYI, `pytest` will pass the same `tmpdir` instance to a test and its fixtures: so the `tmpdir` used by the `paths` fixture will be the same as the `tmpdir` used within the test. I don't know if that changes anything, but it's useful background info regardless.)